src/builder.rs

```diff
@@ -112,32 +102,40 @@ impl AstarteBuilder {
             credentials_secret: credentials_secret.to_owned(),
             pairing_url: pairing_url.to_owned(),
             interfaces: HashMap::new(),
-            build_options: None,
+            mqtt_options: None,
             database: None,
             ignore_ssl_errors: false,
             keepalive: std::time::Duration::from_secs(30),
         }
     }
 
-    pub fn with_database<T: AstarteDatabase + 'static + Sync + Send>(&mut self, database: T) {
+    pub fn database<'a, T: AstarteDatabase + 'static + Sync + Send>(
+        &'a mut self,
+        database: T,
+    ) -> &'a mut Self {
         self.database = Some(Arc::new(database));
+        self
     }
 
     /// Set time after which client should ping the broker
     /// if there is no other data exchange
-    pub fn set_keep_alive(&mut self, duration: std::time::Duration) {
+    pub fn keepalive(&mut self, duration: std::time::Duration) -> &mut Self {
         self.keepalive = duration;
+
+        self
     }
 
-    pub fn ignore_ssl_errors(&mut self) {
+    pub fn ignore_ssl_errors(&mut self) -> &mut Self {
         self.ignore_ssl_errors = true;
+
+        self
     }
 
     /// Add an interface from a json file
-    pub fn add_interface_file(
-        &mut self,
+    pub fn interface_file<'a>(
+        &'a mut self,
         file_path: &Path,
-    ) -> Result<&mut Self, AstarteBuilderError> {
+    ) -> Result<&'a mut Self, AstarteBuilderError> {
```

Comment:
I'm not sure I understand, this method can fail, and returns a Result, I don't see the point of propagating the error in another place?

Reply:
Do you know whether in Rust it is considered a best practice for builders to allow a function chain like `builder.foo().unwrap().bar.unwrap()`? As I was saying, it seems this is not covered in The Rust Book

Reply:
> Do you know whether in Rust it is considered a best practice for builders to allow a function chain like `builder.foo().unwrap().bar().unwrap()`? 

Result is returned when it's necessary, I doubt anyone would care about labeling this. We have the question mark operator, it makes no sense to add the complexity to propagate the error in any other way. Unlike the example in the rust book, the non-terminal methods here have side effects.

Reply:
I am also of the opinion to unwrap everything at the end of the call chain and returning `Result` only in the `build` method.
See https://docs.rs/tokio/1.1.0/tokio/runtime/struct.Builder.html#examples

Reply:
+1 on unwrapping at the end

Reply:
> See https://docs.rs/tokio/1.1.0/tokio/runtime/struct.Builder.html#examples
They don't have side effects on the non terminal methods, we do, that example is irrelevant

Please find another example with side effects on non terminal methods as it's the case here

Reply:
The fact that it's not easy to find examples with side effects on non terminal methods is probably an indication of the fact that side effects should not be performed in non terminal methods (and batched at the build stage instead).
Otherwise, can you show me an example of a builder that returns a `Result` in non terminal methods?